### PR TITLE
Ethantr/feature/add-unavailability-form

### PR DESCRIPTION
**Pull request by ethantr**
## Describe your changes
Amy @anananmianyao and I, @ethantr implemented/redesigned the add unavailability feature. The user can navigate to the new tab by pressing the + button in the Calendar section. When the user has filled out the form, the submit button will be enabled and the user can send their new event to the back end.

This consists of:

- View widget
- ViewModel
- Navigation
- Repository
- Client
- and changes to rest API and the preexisting calendar page.

## Issue ticket number and link
FE-135
FE-137
FE-138

### CODE REVIEW

File: fireapp/lib/data/client/unavailability_form_client.dart [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import 'package:injectable/injectable.dart';
+
+@injectable
+class UnavailabilityFormClient {
+
+  final RestClient restClient;
+  UnavailabilityFormClient(this.restClient);
+
+  // Send an HTTP request to create new unavailability event in the calendar
+  Future<void> createUnavailabilityEvent(int userId,UnavailabilityEventPost newEvent){
```
Comment (emilymclean) [inline fireapp/lib/data/client/unavailability_form_client.dart]: Please format with spaces after commas ",". Ditto to the line below

File: fireapp/lib/domain/models/unavailability/unavailability_time.dart [hunk 2]
```diff
@@ -7,7 +7,7 @@ part 'unavailability_time.g.dart';
 @freezed
 class UnavailabilityTime with _$UnavailabilityTime{
   const factory UnavailabilityTime({
-    required int eventId,
+    required int? eventId,
```
Comment (emilymclean) [inline fireapp/lib/domain/models/unavailability/unavailability_time.dart]: I would add a comment here explaining that eventId is only ever null when creating a new instance (will tell future devs that it is safe to force not-null when getting a server response)

File: fireapp/lib/pages/Calendar/calendarForm.dart [hunk 3]
```diff
@@ -2,19 +2,18 @@
 // ignore: file_names
 // ignore_for_file: no_leading_underscores_for_local_identifiers, use_build_context_synchronously, prefer_typing_uninitialized_variables, file_names
```
Comment (emilymclean) [inline fireapp/lib/pages/Calendar/calendarForm.dart:3]: You shouldn't be touching or editing files in /pages. Only in /presentation (where all the good practice code is, /pages only has legacy, non-functioning views). What do these changes do?

File: fireapp/lib/presentation/unavailability_form/unavailability_form_view_model.dart [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+            start: startDateTime,
+            end: endDateTime,
+            periodicity: periodicity);
+        // Submit and navigate back to Calendar if successful.
+        await _unavailabilityFormRepository.createUnavailabilityEvent(
+            userID, newEvent);
+        _submissionState.add(RequestState.success(null));
+        _navigate.add(const UnavailabilityFormNavigation.calendar());
+      } catch (e, stacktrace) {
+        print(stacktrace);
```
Comment (emilymclean) [inline fireapp/lib/presentation/unavailability_form/unavailability_form_view_model.dart]: Remove this line, logger.e already does this.

File: fireapp/lib/presentation/unavailability_form/unavailability_form_view_model.dart [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+    checkFormValidity();
+  }
+
+  void updateEventTitle(String? value) {
+    titleController.text = value ?? "";
+    checkFormValidity();
+  }
+
+  @override
+  Future<void> dispose() async {
```
Comment (emilymclean) [inline fireapp/lib/presentation/unavailability_form/unavailability_form_view_model.dart:145]: You should also be disposing of streams here

File: fireapp/lib/presentation/unavailability_form/unavailability_form_widget.dart [hunk 6]
```diff
@@ -0,0 +1,173 @@
...
+                            children: [
+                              Text(AppLocalizations.of(context)
+                                      ?.addUnavailabilityButton ??
+                                  "Add Schedule"),
+                              if (data is LoadingRequestState)
+                                SizedBox(
+                                  height: 8,
+                                  width: 8,
+                                  child: CircularProgressIndicator(
+                                      color: (Theme.of(context)
```
Comment (emilymclean) [inline fireapp/lib/presentation/unavailability_form/unavailability_form_widget.dart:83]: Please ensure the indentation is consistent

File: fireapp/lib/presentation/unavailability_form/unavailability_form_widget.dart [hunk 7]
```diff
@@ -0,0 +1,173 @@
...
+                    builder: (context, value) {
+                      return DateFormField(
+                          currentValue: value,
+                          onValueChanged: viewModel.updateStartDate,
+                          decoration:
+                              textFieldStylePositioned(context).copyWith(
+                            prefixIcon: const Icon(Icons.calendar_today),
+                            labelText: AppLocalizations.of(context)
+                                ?.enterUnavailabilityStartDate,
+                          ));
```
Comment (emilymclean) [inline fireapp/lib/presentation/unavailability_form/unavailability_form_widget.dart]: Ensure that closing parenthesis end on the same indentation as their opening counterpart (i.e put the second ")" on a new line). I know this feedback seems nitpicky, but declarative UI can quickly become difficult to read without consistent rules (though yours is very readable, I'm just trying to set up good habits)

**Review by emilymclean — CHANGES_REQUESTED** [review]
Most all of my feedback is minor formatting changes, but I do want to figure out what is happening with calendarFormPage before approving. I'll also update the pipeline for unit tests and make a PR into this branch at some point.

File: fireapp/lib/pages/Calendar/calendarForm.dart (hunk 3 above)
Comment (ethantr) [reply fireapp/lib/pages/Calendar/calendarForm.dart:3]: This was a legacy file for the old Calendar Form, which was linked from the button. The only meaningful changes were the CalendarFormRoute between lines 12 - 17 to change the navigation from the old form (the code below lines 20ish) to our new form. 

We avoided making any new MVVM code for the OG calendar page just to keep it contained for this PR. It does look like there's been some changes to this old file from lines 20 below from when we were first looking at it. These can be ignored and won't do anything.  If it's okay, happy to get rid of this file/bloat code entirely when we start coding the new Calendar View?

Comment (emilymclean) [reply fireapp/lib/pages/Calendar/calendarForm.dart:3]: That sounds good to me. I suspected that was what had been changed, just wanted to double check :)

File: fireapp/lib/presentation/unavailability_form/unavailability_form_widget.dart (hunk 7 above)
Comment (ethantr) [reply fireapp/lib/presentation/unavailability_form/unavailability_form_widget.dart]: I agree. I've fixed it now, but I've found out that styling was coming from refactoring with "dart format". Do you know how to change that styling option?

Comment (emilymclean) [reply fireapp/lib/presentation/unavailability_form/unavailability_form_widget.dart]: Dang, didn't realise that. There is literally no way to change dart's default styling settings

**Review by emilymclean — APPROVED** [review]

**Pull request merged**